internal/gcsx/multi_range_downloader_wrapper.go

```diff
@@ -105,37 +116,98 @@ func (mrdWrapper *MultiRangeDownloaderWrapper) GetRefCount() int {
+	if wasZero && mrdWrapper.mrdCache != nil {
+		deletedEntry := mrdWrapper.mrdCache.Erase(wrapperKey(mrdWrapper))
+		if deletedEntry != nil {
+			logger.Tracef("MRDWrapper (%s) erased from cache", mrdWrapper.object.Name)
+		}
+	}
 }
 
-// DecrementRefCount decrements the refcount. In case refcount reaches 0, cleanup the MRD.
+// DecrementRefCount decrements the refcount. When refCount reaches 0, the wrapper
+// with added to the cache for potential reuse. In case, cache is not enabled, MRD
+// is closed immediately.
 // Returns error on invalid usage.
 // This method should be called exactly once per user of this wrapper
 // when MultiRangeDownloader is no longer needed & can be cleaned up.
 func (mrdWrapper *MultiRangeDownloaderWrapper) DecrementRefCount() (err error) {
 	mrdWrapper.mu.Lock()
-	defer mrdWrapper.mu.Unlock()
 
 	if mrdWrapper.refCount <= 0 {
+		mrdWrapper.mu.Unlock()
 		err = fmt.Errorf("MultiRangeDownloaderWrapper DecrementRefCount: Refcount cannot be negative")
 		return
 	}
 
 	mrdWrapper.refCount--
-	if mrdWrapper.refCount == 0 && mrdWrapper.Wrapped != nil {
-		mrdWrapper.handle = mrdWrapper.Wrapped.GetHandle()
-		mrdWrapper.Wrapped.Close()
-		mrdWrapper.Wrapped = nil
+	// Do nothing if MRD is in use.
+	if mrdWrapper.refCount > 0 {
+		mrdWrapper.mu.Unlock()
+		return
+	}
+
+	// No cache: close MRD immediately
+	if mrdWrapper.mrdCache == nil {
+		mrdWrapper.closeLocked()
+		mrdWrapper.mu.Unlock()
+		return
+	}
+
+	// Cache: add the wrapper to cache and evict overflow wrappers.
+	evictedValues, err := mrdWrapper.mrdCache.Insert(wrapperKey(mrdWrapper), mrdWrapper)
+	if err != nil {
+		logger.Errorf("failed to insert wrapper (%s) into cache: %v", mrdWrapper.object.Name, err)
+		mrdWrapper.mu.Unlock()
+		return
+	}
+	logger.Tracef("MRDWrapper (%s) added wrapper to cache", mrdWrapper.object.Name)
+
+	// Evict outside all locks to avoid deadlock.
+	mrdWrapper.mu.Unlock()
+	for _, wrapper := range evictedValues {
+		mrdWrapper, ok := wrapper.(*MultiRangeDownloaderWrapper)
+		if !ok {
+			return fmt.Errorf("invalid value type during cache eviction")
+		}
+		mrdWrapper.CloseMRDForEviction()
+	}
+	return nil
+}
+
+// CloseMRDForEviction closes the MRD when evicted from cache.
+// This method is called after wrapper was removed from cache for eviction.
+// Race protection: wrapper could be reopened (refCount>0) or re-added to cache before eviction.
+func (mrdWrapper *MultiRangeDownloaderWrapper) CloseMRDForEviction() {
+	mrdWrapper.mu.Lock()
+	defer mrdWrapper.mu.Unlock()
+
+	// Check if wrapper was reopened (refCount>0) - must skip eviction.
+	if mrdWrapper.refCount > 0 {
+		return
+	}
+
+	// Check if wrapper was re-added to cache (refCount went 0→1→0 in between eviction and closure.)
+	// Lock order: wrapper.mu -> cache.mu (consistent with Increment/DecrementRefCount)
+	if mrdWrapper.mrdCache != nil && mrdWrapper.mrdCache.LookUpWithoutChangingOrder(wrapperKey(mrdWrapper)) != nil {
+		return
+	}
+
+	if mrdWrapper.closeLocked() {
```

Comment:
Are you returning a bool just for logging purpose? you can just add a log in closeLocked to say that we are closing it.

Reply:
Given we can removed the immediate closure logic when cache is disabled, we can move the log inside closeLocked() only.